sympy/combinatorics/pc_groups.py

```diff
@@ -0,0 +1,239 @@
+from sympy.core import Basic
+from sympy import sieve
+from sympy.combinatorics.permutations import Permutation
+from sympy.combinatorics.perm_groups import PermutationGroup
+
+class PcGroup(Basic):
+
+    is_group = True
+    is_solvable = True
+
+    def __init__(self, _pcgs):
+        self.perm_group = PermutationGroup(_pcgs)
+        self.pc_series = self._pc_series()
+        self.pcgs = self._compute_pcgs()
+
+    def _pc_series(self):
+        return self.perm_group.composition_series()
+
+    def _compute_pcgs(self):
+        # computes the generating sequence for polycyclic groups.
+        series = self.pc_series
+        pcgs = []
+        for i in range(len(series)-1):
+            for g in series[i].generators:
+                if not g in series[i+1]:
+                    pcgs.append(g)
+        return pcgs
+
+    def relative_orders(self):
+        rel_orders = []
+        for i in range(len(self.pc_series)-1):
+            G = self.pc_series[i]
+            H = self.pc_series[i+1]
+            rel_orders.append(G.order()//H.order())
+        return rel_orders
+
+    def is_prime_order(self):
+        for order in self.relative_orders():
+            if order not in sieve:
+                return False
+        return True
+
+    def length(self):
+        return len(self.pcgs)
+
+    def pc_element_exponent(self, element):
+        series = self.pc_series
+        pcgs = self.pcgs
+        exponent = [0]*len(series)
+        for i in range(len(series)):
+            exp = 0
+            if not element in series[i]:
+                for j in range(len(pcgs)):
+                    element = (pcgs[j]**-1)*element
+                    exp = exp + 1
+                    if element in series[i]:
+                        exponent[i] = exp
+                        break
+        return exponent
+
+
+def minimal_uncollected_subwords(word, relative_order):
+    """
+    Returns the minimal uncollected subwords.
+
+    Examples
+    ========
+    >>> from sympy.combinatorics.pc_groups import minimal_uncollected_subwords
+    >>> from sympy.combinatorics.free_groups import free_group
+
+    Example 8.7 Pg. 281 from Handbook
+    >>> F, x1, x2 = free_group("x1, x2")
+    >>> word = x2**2*x1**7
+    >>> relative_order = [3, 2]
+    >>> minimal_uncollected_subwords(word, relative_order))
+    {x1*x2**2: 0, x1**7: 1}
+
+    """
+    array = word.array_form
+    uncollected_subwords = {}
+    for i in range(len(array)-1):
+        if array[i+1][1] > 0:
+            # case-1:  v = x[i]**a*x[i+1]
+            uncollected_subwords[array[i][0]**array[i][1]*array[i+1][0]] = 0
```

Comment:
I think that `x[i]` and `x[i+1]` should be group elements.

Reply:
Do you mean converting them into group element using `dtype()`?

Reply:
I think that there is a method for the conversion.

Reply:
I'm not able to find any such method!

Reply:
It seems that I was thinking of `letter_form_elm` that is used in `rewritingsystem`. It calls `dtype` internally but apparently cannot be applied here, at least not easily. So it looks like `dtype` has to be used more explicitly for the conversion. Anyway, some new code has to be introduced to manage the ordering of the generators.